api/src/main/java/org/pmiops/workbench/reporting/ReportingSnapshotServiceImpl.java

```diff
@@ -1,59 +1,100 @@
+  private static final Logger log = Logger.getLogger(ReportingSnapshotServiceImpl.class.getName());
 
   private final Clock clock;
   private final ReportingMapper reportingMapper;
-  private Random random;
+  private final Random random;
+  private final Provider<Stopwatch> stopwatchProvider;
   private final UserService userService;
   private final WorkspaceService workspaceService;
 
+  // Define immutable value class to hold results of queries within a transaction. Mapping to
+  // Reporting DTO classes will happen outside the transaction.
+  private static class EntityBundle {
+    private final List<DbUser> users;
+    private final List<DbWorkspace> workspaces;
+
+    public EntityBundle(List<DbUser> users, List<DbWorkspace> workspaces) {
+      this.users = users;
+      this.workspaces = workspaces;
+    }
+
+    public List<DbUser> getUsers() {
+      return users;
+    }
+
+    public List<DbWorkspace> getWorkspaces() {
+      return workspaces;
+    }
+  }
+
   public ReportingSnapshotServiceImpl(
       Clock clock,
       ReportingMapper reportingMapper,
       Random random,
+      Provider<Stopwatch> stopwatchProvider,
       UserService userService,
       WorkspaceService workspaceService) {
     this.clock = clock;
     this.reportingMapper = reportingMapper;
     this.random = random;
+    this.stopwatchProvider = stopwatchProvider;
     this.userService = userService;
     this.workspaceService = workspaceService;
   }
 
   @Override
   public ReportingSnapshot takeSnapshot() {
-    return new ReportingSnapshot()
-        .captureTimestamp(clock.millis())
-        .researchers(getResearchers())
-        .workspaces(getWorkspaces());
-  }
+    final EntityBundle entityBundle = getApplicationDbData();
+    final Stopwatch stopwatch = stopwatchProvider.get().start();
 
-  private List<ReportingResearcher> getResearchers() {
-    final List<DbUser> users = userService.getAllUsers();
-    return reportingMapper.toReportingResearcherList(users);
+    final List<ReportingWorkspace> workspaces =
+        reportingMapper.toReportingWorkspaceList(entityBundle.getWorkspaces());
+    for (ReportingWorkspace model : workspaces) {
+      model.setFakeSize(
+          getFakeSize()); // TODO(jaycarlton): remove after initial query & view testing
+    }
+
+    final ReportingSnapshot result =
+        new ReportingSnapshot()
+            .captureTimestamp(clock.millis())
+            .researchers(reportingMapper.toReportingResearcherList(entityBundle.getUsers()))
+            .workspaces(workspaces);
+    stopwatch.stop();
+    log.info(LogFormatters.duration("Conversion to ReportingSnapshot", stopwatch.elapsed()));
+    return result;
   }
 
-  private List<ReportingWorkspace> getWorkspaces() {
+  // Retrieve all the data we need from the MySQL database in a single transaction for
+  // consistency.
+  @Transactional(readOnly = true)
+  protected EntityBundle getApplicationDbData() {
```

Comment:
Per Spring documentation: 
Method visibility and @Transactional

When using proxies, you should apply the @Transactional annotation only to methods with public visibility. If you do annotate protected, private or package-visible methods with the @Transactional annotation, no error is raised, but the annotated method does not exhibit the configured transactional settings. 

The annotation needs to exist on method called from outside the proxy class. So the `takeSnapshot` method would be a candidate for the annotation.

Reply:
Oh ok. I was hoping to avoid the time to convert the entries extending the transaction any longer, but I bet that's negligible in practice.